jobrunner/job_executor.py

```diff
@@ -0,0 +1,114 @@
+Study = Tuple[str, str]  # Git repo and commit
+
+
+@dataclass
+class JobDefinition:
+    study: Study  # the study defining the action for this job
+    workspace: str  # the workspace to run the job in
+    action: str  # the name of the action that the job is running
+    image: str  # the Docker image to run
+    args: List[str]  # the arguments to pass to the Docker container
+    env: Mapping[str, str]  # the environment variables to set for the Docker container
+    inputs: InputSpec  # the files that the job requires, a mapping of file paths to the actions that produced them
+    outputs: OutputSpec  # a description of the expected outputs (privacy level mapped to patterns)
+    allow_database_access: bool  # whether this job should have access to the database
+
+
+@dataclass
+class JobResults:
+    state: State
+    status_code: Optional[StatusCode]
+    status_message: str
+    outputs: Mapping[str, str]
+    unmatched_outputs: List[str]
+
+
+class JobAPI(Protocol):
+    def run(self, job_id: str, definition: JobDefinition) -> None:
+        """
+        Run a job.
+
+        This method must be idempotent; it may be called more than once with the same job_id, in which case only one
+        job should be created. It must also be idempotent in the face of errors; if it throws an exception because
+        job creation has failed due to a transient error it may be called again to retry the operation and this
+        should succeed if possible. (The implementation may provide a configuration option which breaks this
+        idempotency by preserving resources after a failure to aid debugging.)
+
+        The specified image must be run, with the provided arguments and environment variables. The implementation
+        may add environment variables to those in the job definition as necessary for the backend.
+
+        The job should be run without any network access, unless definition.allow_database_access is set to True,
+        in which case it should be run with a network allowing access to the database and any configuration needed to
+        contact and authenticate with the database should be provided as environment variables.
+
+        The job must be run with a workspace directory at /workspace in the filesystem (this is expected to be a
+        volume mounted into the container, but other implementations are permitted). The workspace must contain a
+        checkout of the study and any inputs specified in the job definition, copied from long-term storage.
+
+        When a job finishes, whether or not it has succeeded, any output files that match the output spec in the
+        definition must be copied up to long-term storage, both to the high privacy workspace and, if called for in
+        the spec, to the medium privacy workspace. A non-disclosive log file containing useful metadata about the job
+        should also be written to both workspaces in long-term storage.
```

Comment:
This doesn't sound quite right to me. I think we want to say something like:

> Anything written to stdout/stderr by the job should be captured and written to a file named `metadata/{action_name}.log` in both the high and medium privacy workspaces. (Ideally, with each line prefixed by timestamps.)

There's a separate point here which is useful for debugging and audit but I suppose doesn't strictly form part of the contract. That's writing the above log file and the "big bag of useful metadata" into a separate log directory in persistent storage, named using the job_id. I'm not sure where these kind of nice-to-haves are best documented. Or maybe this is more than just a nice-to-have, I'm not sure.

Reply:
I'd lean towards more than just nice to have, to be honest, as I think debugging anything will be nigh on impossible without it

Reply:
> named using the job_id

I don't see this anywhere. Looks like the filename has the action?

Reply:
Oh no, I've found it now. Thanks.

Reply:
Yes, it's not done in the most transparent way!

The log and metadata dump happens here:
https://github.com/opensafely-core/job-runner/blob/2214002be42ac478a8111a42901cd179cb4de65c/jobrunner/manage_jobs.py#L296-L300

And the directory it's dumped into uses the container name, which contains the job_id (along with some human readable stuff):
https://github.com/opensafely-core/job-runner/blob/2214002be42ac478a8111a42901cd179cb4de65c/jobrunner/manage_jobs.py#L481-L484

I'm not suggesting that specific scheme needs to be replicated. But some way of getting logs and metadata for historical jobs is very useful, and they're small enough that we can effectively keep them forever.

Reply:
@evansd I've rewritten this. I'm trying to make it both clear and correct. I've also taken into account the discussion we had about the relationship between the worksapce and the privacy levels in that call just now.

(The explanation here isn't intended to be completely standalone. I assume there will be a separate doc, covered in another card, which explains the workspace concept and how long-term storage and privacy levels work.)

Could you take another look, please?

Reply:
I think I'm exactly mirroring the current implementation apart from not specifying the appending of a subset of the metadata to the log file, which feels a bit fiddly and may be optional. But I can add it too if you think we really need it.

Reply:
This looks good to me. I agree about not appending anything to the logs. That's too fiddly and I think we can do everything we need using the entrypoint script in a much cleaner way.